### PR TITLE
[SPARK-11849][SQL] Analyzer should replace current_date and current_timestamp with literals

### PR DESCRIPTION
**Pull request by rxin**
We currently rely on the optimizer's constant folding to replace current_timestamp and current_date. However, this can still result in different values for different instances of current_timestamp/current_date if the optimizer is not running fast enough.

A better solution is to replace these functions in the analyzer in one shot.

### CODE REVIEW

**Comment by rxin** [comment]
cc @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 1]
```diff
@@ -218,4 +219,39 @@ class AnalysisSuite extends AnalysisTest {
...
+  test("analyzer should replace current_timestamp with literals") {
+    val in = Project(Seq(Alias(CurrentTimestamp(), "a")(), Alias(CurrentTimestamp(), "b")()),
+      LocalRelation())
+
+    val min = System.currentTimeMillis() * 1000
+    val plan = in.analyze.asInstanceOf[Project]
+    val max = (System.currentTimeMillis() + 1) * 1000
+
+    val lits = new scala.collection.mutable.ArrayBuffer[Long]
+    plan.transformAllExpressions{ case e: Literal =>
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: nit: space before `{`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 2]
```diff
@@ -218,4 +219,39 @@ class AnalysisSuite extends AnalysisTest {
...
+      lits += e.value.asInstanceOf[Long]
+      e
+    }
+    assert(lits.size == 2)
+    assert(lits(0) >= min && lits(0) <= max)
+    assert(lits(1) >= min && lits(1) <= max)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: should we also assert that `lits(0) == lits(1)`?

**Comment by rxin** [comment]
k updated.

**Comment by cloud-fan** [comment]
LGTM

**Comment by rxin** [comment]
Merging this. Thanks.

**Pull request closed without merging**